plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```
@@ -0,0 +1,206 @@
+            }
+            p.pending_snapshots->emplace_back(si);       
+         });
+      }
+   }
+
+   // snapshot executor
+   void set_create_snapshot_fn(std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> fn) {
+      _create_snapshot = std::move(fn);
+   }
+
+   void execute_snapshot(uint32_t srid) {
+      _inflight_sid = srid;
+      auto next = [srid, this](const std::variant<fc::exception_ptr, producer_plugin::snapshot_information>& result) {
+         if(std::holds_alternative<fc::exception_ptr>(result)) {
+            try {
+               std::get<fc::exception_ptr>(result)->dynamic_rethrow_exception();
+            } catch(const fc::exception& e) {
+               elog( "snapshot creation error: ${details}", ("details",e.to_detail_string()) );
+               appbase::app().quit();
+            } catch(const std::exception& e) {
+               elog( "snapshot creation error: ${details}", ("details",e.what()) );
+               appbase::app().quit();
+            }
+         } else {
+            // success, snapshot finalized
+            auto snapshot_info   = std::get<producer_plugin::snapshot_information>(result);
+            auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+            auto  snapshot_req   = snapshot_by_id.find(srid);
+
+            if (snapshot_req != snapshot_by_id.end()) {               
+               if (auto pending = snapshot_req->pending_snapshots; pending) {                 
+                  auto it = std::remove_if(pending->begin(), pending->end(), [&snapshot_info](const producer_plugin::snapshot_information & s){ return s.head_block_num <=  snapshot_info.head_block_num; });
+                  pending->erase(it, pending->end());
+                  _snapshot_requests.modify(snapshot_req, [&pending](auto& p) { 
+                     p.pending_snapshots = std::move(pending);
+                  });
+               }
+            }
```

Do you want to report an error if `snapshot_req` was not found?

---

done

---

not sure if we should gracefully exit here, probably not. un-finalized snapshot (should not happen) does not sound like terminal case.